### ISSUE TITLE
DID URLs for CredDefs referencing different schema Objects/versions

### ISSUE THREAD

**Issue by domwoe** [hyperledger/indy-did-method#38]
I had another look at the [DID URLs for Indy Object Identifiers](https://hyperledger.github.io/indy-did-method/#did-urls-for-indy-object-identifiers) section.

1. Was the notation in the schema section on purpose? Currently, the indy nomenclature uses `version` for the last part of the identifier. In the spec, `tag` is used.

2. Identifying legacy Cred Defs (and therefore Rev Regs) could be ambiguous. In the new DID URL style identification we only specify the DID and a name. The name is the `tag` in the current indy nomenclature.  I just looked up a few Cred Defs on the IDunion ledger and some use the tag for versioning, i.e there might be Cred Defs with the same DID and "name" (e.g. 1.0) corresponding to different Schemas. I have not looked at how common this is and would expect that it is not very common since smartphone wallets typically use the tag/name as the display name for credentials. Reading the last sentence: _Specifically, the same named claim def can no longer be associated with different [SCHEMA](https://hyperledger.github.io/indy-did-method/#term:schema) objects, you are aware of the issue but I guess we should at least define what happens if there is ambiguity in current ledger objects that existed before we enforce did indy rules.

**Comment by swcurran** [comment]
The "tag" reference is indeed in correct, and should be version.

I agree we need some clarification about the schema object to which the ClaimDef applies. this relates to the question we have of referencing a schema object on another ledger.

**Comment by domwoe** [comment]
To get an idea of the issue I've checked the sovrin main net for credential definitions from a DID with duplicate tags. There are 205 occurrences in which the current Cred Def DIDURL would be ambiguous. 

One example would be [Gs6cQcvrtWoZKsbBhD3dQJ](https://indyscan.io/txs/SOVRIN_MAINNET/domain?page=1&pageSize=50&filterTxNames=[%22CLAIM_DEF%22]&sortFromRecent=true&search=Gs6cQcvrtWoZKsbBhD3dQJ).
There are 4 Cred Defs associated with this DID. All Cred Defs reference a distinct schema but share the same tag/name.

```
{
   "XGjatCu9mSLUWiJznui3CF":{
      "cds":2
   },
   "Gs6cQcvrtWoZKsbBhD3dQJ":{
      "mpc":4
   },
   "Y6LRXGU3ZCpm7yzjVRSaGu":{
      "DemoBasicIdentity":2
   },
   "12Umxoybo8Lht7w8AiH6m6":{
      "tag1":2
   },
   "T21MWBVwD6jaVzcnmmGhFK":{
      "tag1":2
   },
   "PrKeQGzzRjuNLjciqV6FbA":{
      "84ccc5e4-692e-475c-8230-19a1b343d850":4
   },
   "KE4CCjJWqjFQtyGeix3C14":{
      "3d2abbc6-66d3-435e-a543-3fa5a3ece39a":4
   },
   "SenJJCrEZRMf4WJoHvuPwH":{
      "bb97f7a8-6e7f-4f81-80f1-af943e4ba053":4
   },
   "JSVdVn5x7UE4tcqAP9qv1x":{
      "zada":2
   },
   "KtqL7bYAME1ozArKB8hZhU":{
      "Prueba":3
   },
   "74RKMKkqSVJ6xwUMQxaSeU":{
      "ff5dff10-190f-4c18-ad2e-a22a87119cad":4
   },
   "8EkP4VHv8qxYkfJq7K9CBc":{
      "tag1":2
   },
   "RwSuTfckvgWjXTXwiTVFtP":{
      "latest":5
   },
   "CbgReDvw4jCPCpJQzb6REE":{
      "38decafa-55ae-4a32-bcbe-2a3f744de878":4
   },
   "2E9s1wF1nFzMDBkU4f8DLs":{
      "tag1":2
   },
   "DPCXnM6cfxy5GUeL72pkvN":{
      "af28f1b1-493c-4b9a-a482-f8529bb15297":5
   },
   "FLfpAM9gWWLBZ4Lq6iGE9F":{
      "3c939071-bd9d-4059-91fb-f7206af2fbce":8
   },
   "SPmxVt8ZX3vKo7LJx62ToL":{
      "7a794f30-7a8c-4114-a833-de538b53de41":5
   },
   "FRqVJhJecsgke3s1G2QUcQ":{
      "5b12a1c2-7064-4326-90e5-d9cf0b42dafb":5
   },
   "4sKco14kTsS6WopBpDmmtU":{
      "c47a20c1-05a3-4ba3-9546-b9316a8d5343":5
   },
   "Bpd4Pt7KsZAPVJRVau8eeh":{
      "b058083d-ca48-40f2-b9ee-b2c5714e1b0c":5
   },
   "PJ5kqkasKJ8D4NUWRedSGo":{
      "324827fa-24a1-4ce7-958c-addae5255576":9
   },
   "5WpV3ojCQAChXSVTURef8X":{
      "0484e982-4034-4b27-b2cf-a637d8904b82":7
   },
   "KH4nvYj9gbuPGMD1Vqvi8W":{
      "31cfda10-11ef-430c-8a98-88e547a0e7ed":11
   },
   "7dKAPKLu5RrAq5iXD5Mmca":{
      "5f985e88-a7dc-4c31-bf99-830c0f22692f":7
   },
   "MiG7io5NhrXzGGU8CRezxv":{
      "b62708c9-2c11-45c3-8d4c-7a57bdc21c77":7
   },
   "ixQy9iZqyNrPZZzNP7bpW":{
      "e187df8d-4145-4515-9450-5e11820b1d74":4
   },
   "GiTG6YWQEiAf3wSgpetK1P":{
      "b088b44a-871e-4639-81ad-4ccd38d228c9":6
   },
   "UZ3VMUuovsr63R3qnz7NoB":{
      "9b4ef9c0-3d94-4baa-95ac-448064e15d9f":6
   },
   "2jmjkGVzRGe9TLWCmM4BYX":{
      "c29f0c8d-8d55-41e2-bc0f-3a3fd048618b":4
   },
   "R1ABYMXJL1CDU1S8Rrbc1y":{
      "d2632e66-1a86-447e-b005-35d2ad376498":6
   },
   "4YEYRTQhgUY1NiL7M3jz6m":{
      "ae22eb36-8f2e-434e-81f7-31e5c72e5d4f":16
   },
   "NVHn7ohtGt42Z1QrfMhRS9":{
      "ddcbb8a5-a762-4244-93b9-33e35bc0c101":9
   },
   "VC3Gx1HoAz8iYVcfY47gjJ":{
      "hospital_affiliation":2
   },
   "3pRGFwNi74rhnpGF4Qw7Nn":{
      "default":4
   },
   "2jNK7j4uqPvLCJ2r2wo8RN":{
      "6fa7cbfa-3bd1-42e8-bcb5-5f757da30e29":4
   },
   "FqtdLvcM2i4Q7Ewd8Uf8FJ":{
      "L2_DOC":2,
      "L1_DOC":2,
      "L2":2,
      "L1":2,
      "Member Back2Work":2
   },
   "CPLt8bMQ2VwUv6kDRkekYe":{
      "latest":2
   },
   "RkayAfMr1FzjcK89NX8NtM":{
      "latest":3
   },
   "D2yuJgtkJjDVveH6ihRscA":{
      "tag1":2
   },
   "RYqWk33jqZoEqsvDxnLvmC":{
      "tag1":2
   },
   "LUBD71fHzZKnHjoTfaDiMb":{
      "tag1":2
   },
   "6BrHyg7SoMRmbgQjqJrkcH":{
      "a94fdbc5-937c-4233-bdfc-5909300dfff4":15
   },
   "RDmfeHMBEy7w8AQ7KXyGNi":{
      "default":6
   },
   "9rE9T4VsG5bBUC1EhbrkZG":{
      "tag1":2
   },
   "Tzmf8z4UrFec6RGMEnH3F5":{
      "latest":37
   },
   "QosCTowZAZxYU5RSeMh9UJ":{
      "Default":2
   },
   "59C8n5ph5tTGixr4sUrQCD":{
      "tag1":2
   },
   "Stkccdro6kh5ie5Ek7nqcL":{
      "tag1":2
   },
   "RG6fJfmLLJ228FG5TYEbZf":{
      "prod":5,
      "tag1":2
   },
   "7dCZLN3via4S1eMjsF4WPB":{
      "tag1":2
   },
   "DAsJrUBcQEDMEGY5UkhMqY":{
      "latest":3
   },
   "5sVrpQzHx31fiFmgAQLdjs":{
      "tag1":2
   },
   "EfPBZtcwxxn3eD2nrVUbUo":{
      "tag1":2
   },
   "PkBzxCwZr2S2MfmoxjgMAY":{
      "stage":2
   },
   "8KusBegNmSG2bitSnf14tE":{
      "stage":9
   },
   "AMRGUn9g8eUEExjmR3xB51":{
      "Default":3
   },
   "T4ty9vvEbYRFVGzufeEQ2t":{
      "tag1":2
   },
   "2qnry2qxgP7RdmrPG3rSzm":{
      "tag1":2
   },
   "Hrsqn12pP2okHccjJqUS9":{
      "Default":2
   },
   "C2b4EVBuFn1kcp7Riz4fEk":{
      "tag1":2
   },
   "PHfUSDmdG2j2XCkyFE6wwh":{
      "latest":3
   },
   "QtPVoEqgC5AVMwwKgixzvV":{
      "latest":2
   },
   "K9VwgzhME97RVqLfLY88oH":{
      "latest":2
   },
   "2uTPgRPNzcpWXynW2funSj":{
      "latest":2
   },
   "WqucmizfWtDJ5vda11sJgJ":{
      "latest":2
   },
   "W2u9PzjDmhKM5xLABjAqav":{
      "latest":4
   },
   "PJPdgJuKKiEmwd8jgquw4S":{
      "latest":2
   },
   "G8G4n8xxNYnUMfHaxkZNhE":{
      "tag1":2
   },
   "NGYKKkDqYn2YfUYcjhArc5":{
      "tag1":2
   },
   "MYKzh53hmgkdXM1MSgbUQV":{
      "tag1":2
   },
   "G7dkz6hRouuYzn573bJpJF":{
      "tag1":2
   },
   "12nQeyzhkpckWgep2hgZVw":{
      "tag1":2
   },
   "4xE68b6S5VRFrKMMG1U95M":{
      "default":4
   },
   "JCrQh7urB68mgmxLxpUdHP":{
      "tag1":2
   },
   "26z6BEx9L3f3Ppu1Ukzd51":{
      "tag1":2
   },
   "Vhc1G3TXhwm3E7Z57MohmW":{
      "tag1":2
   },
   "4SnhHSYPgV4ZxFwFSGF7LN":{
      "tag1":2
   },
   "HUzVAJ6ESFU1WbFjTs5vCx":{
      "tag1":2
   },
   "A7E9dAagzs3w217B4pSZZm":{
      "Impfnachweis":7
   },
   "NjMtuygSs5EEvsTsaQTRR7":{
      "tag1":2
   },
   "FBGCSJugXiG1f48jSQsqut":{
      "tag1":2
   },
   "3V1AqDs1XtcdXhPkJGC5pw":{
      "tag1":2
   },
   "CBZzse5iq9Wg9goBQGCe4i":{
      "tag1":2
   },
   "Xfzn5GyRg1AdP4FwBDEcJh":{
      "tag1":2
   },
   "9fjdNEJ22dDu7Y2fmRw8uo":{
      "tag1":2
   },
   "YauDaoqttsxb66r7cp93CM":{
      "tag1":2
   },
   "NrhhiX3yR9RhH5zkKkqnSU":{
      "tag1":2
   },
   "6nqZCMzS9GQ5FQwhHv7cQp":{
      "tag1":2
   },
   "2PsqtgEz8jVwYPAMj8pyuD":{
      "tag1":2
   },
   "ACLoJxe923SwYv1v34BYWw":{
      "tag1":2
   },
   "ToGeB7gWi4yoPepUzpEYd1":{
      "tag1":2
   },
   "8T2s6PV6mypom1Q2ovrCy":{
      "tag1":2
   },
   "NZGg9K8Bhi2rBPk529YRcz":{
      "tag1":2
   },
   "MaCn3bfFgBe4mtumsCwaJo":{
      "tag1":2
   },
   "V62UWVF73wTcjQMxqmArsj":{
      "tag1":2
   },
   "DNsnxGMxNkuMc8wifSvJzh":{
      "tag1":2
   },
   "BeyBw7ZevdBKwXhbcJCuLz":{
      "tag1":2
   },
   "UDExwMXij2CPFtWNErcGWU":{
      "tag1":2
   },
   "SiLndbq78ZddLbsCv5sZxL":{
      "tag1":2
   },
   "PTQSzgiDw65Xsmr9vYzCoK":{
      "tag1":2
   },
   "WaL6XTgwhTaswgrwRXrdpM":{
      "tag1":2
   },
   "AGMwFvEZrxLErfKP82HRBD":{
      "tag1":2
   },
   "TTWdfVNdHxCBtwqcrzZDgm":{
      "tag1":2
   },
   "Q5miwN5sArEfNqD7azTXAt":{
      "tag1":2
   },
   "8DyS4krAKdRQiZq5Wo3v7Q":{
      "tag1":2
   },
   "Rg5zNQP9rCaKHd3RGqMHs3":{
      "tag1":2
   },
   "EAFtNK4xNi9pgQWpayiCTS":{
      "tag1":2
   },
   "2vAgJBto6TLVA5uftwfyB4":{
      "tag1":2
   },
   "G1i4NqPmKovcFSJp66hs8k":{
      "tag1":2
   },
   "ACrsTqgC8kqBhe4bRuMk8":{
      "tag1":2
   },
   "NefJNo2LXGX6W2Jyh1Suv7":{
      "tag1":2
   },
   "8yfaaV4T1Y2Lea3ZywKjMS":{
      "tag1":2
   },
   "21ibSa5YFVkF6G1j18ktHS":{
      "tag1":2
   },
   "Y9TavPJ3wwCUpoPxPrfTW8":{
      "tag1":2
   },
   "7cu6GBzGCxqgho4Xq8MsaW":{
      "tag1":2
   },
   "xBMJxuN9Z5PvJ4FJiUcN6":{
      "tag1":2
   },
   "Qsd8dCmqEAVhrR3KuAGrrV":{
      "tag1":2
   },
   "F4FZvYfSQ16z88zDX9Cowp":{
      "tag1":2
   },
   "WDoKihJuYYxSJXK21mZz1o":{
      "tag1":2
   },
   "GPYhByfPPXA2TXXApE1pdo":{
      "tag1":2
   },
   "7NdMzpGUBPcSnadraBENHV":{
      "tag1":2
   },
   "9fn2S3EdXR4zEgTvA9SZzg":{
      "tag1":2
   },
   "42nNSvYGdswJgwV8o1Joqg":{
      "tag1":2
   },
   "KJy4brY6c13gM7J7BYaoug":{
      "tag1":2
   },
   "JFiusFBwvpnEtTVsfSnVgY":{
      "tag1":2
   },
   "Mnq9mP49P6cBLGKgxi1Pm3":{
      "tag1":2
   },
   "WAxXoMmuh1hDndkQ9EuWp":{
      "tag1":2
   },
   "937dAQr76MQTagbzDtwQUZ":{
      "tag1":2
   },
   "5tEhymeLnEi3DtrBaT4JLp":{
      "tag1":2
   },
   "GZ4gF8L786KZS6PGKMnGrd":{
      "tag1":2
   },
   "EPotyVh44SpNjPta6RRabs":{
      "tag1":2
   },
   "VEoLwZf2NpLyJEe1qnqvd1":{
      "tag1":2
   },
   "HxipYRVUjW93xEUVouz8ij":{
      "tag1":2
   },
   "KahJuAzbnuVSSd6LrePs8x":{
      "tag1":2
   },
   "LZzRQSoonqe1hhefc8kLj2":{
      "tag1":2
   },
   "AWY4BHqoKVKtjRDbcf2ScU":{
      "tag1":2
   },
   "UmkmLRNTw2t2GSLaZMoThJ":{
      "tag1":2
   },
   "R3siv7XQFwVYfuW7JhKoen":{
      "tag1":2
   },
   "8R3SdfanT2bEJMwy1Q1PQL":{
      "tag1":2
   },
   "WuDUppgob2xnGbZMkNCqtA":{
      "tag1":2
   },
   "VcxrA3ox4SMb8zfHvZeeva":{
      "tag1":2
   },
   "JMTKxP74kefFazen16f8st":{
      "tag1":2
   },
   "26KgwHDyPtxmvwrDdRkqet":{
      "tag1":2
   },
   "WGpBJ9pZpDK8QgXrgRU5Mf":{
      "tag1":2
   },
   "L8RHBzddunmqGE1HG1xYwD":{
      "tag1":2
   },
   "3eURP59ufxZmPoR4z7sPQ7":{
      "tag1":2
   },
   "6g1A4fxFiUf8LTyM5Tqgkv":{
      "tag1":2
   },
   "DfW4ZgFr54E8zWKDz4SxTW":{
      "tag1":2
   },
   "34qdX2WQxr8D6KFpT2n8TK":{
      "tag1":2
   },
   "5mq2pEv5EHtn3h5XVW6f8s":{
      "tag1":2
   },
   "G4XVLVqvLrFmLZGCNrPMcY":{
      "tag1":2
   },
   "RqXVSrUkz5umkQVEKKNGP1":{
      "tag1":2
   },
   "VMcJAnHabj1fuVnqEewN9B":{
      "tag1":2
   },
   "7k25fH9oR4JjfESifsjMa3":{
      "tag1":2
   },
   "PBnzTevFWPjRtg6ThAMqQU":{
      "tag1":2
   },
   "XtUUvVCica8N5pthe3GDx2":{
      "tag1":2
   },
   "WvQbn4afTX5P2xnDAtrVA6":{
      "tag1":2
   },
   "A6QSHXJbuwbP4GBMGygL5q":{
      "tag1":2
   },
   "HDoy3hN1e7DrKKycQrchcZ":{
      "tag1":2
   },
   "AUN9QuPXUJybRaeD7QnKYa":{
      "tag1":2
   },
   "NTMYQjK8XAz1WuHbsAp9zK":{
      "tag1":2
   },
   "7vFksWRnLQbHCevVchN3GM":{
      "tag1":2
   },
   "KhUg9rzWjxo6aWzDLTvTxF":{
      "tag1":2
   },
   "X52HKkC2LcY4Cjeacm3bnm":{
      "tag1":2
   },
   "YT7D2ZcsCi6j8S4d41rhDN":{
      "tag1":2
   },
   "F2XXAb3yaQesBVQyudHy7M":{
      "tag1":2
   },
   "NUub1Ut74ZQyhkEidGVZaK":{
      "tag1":2
   },
   "Gk1NPjxdiRSMDt6br2CgVT":{
      "tag1":2
   },
   "JwqWgVaftmppEqxfMhciWk":{
      "tag1":2
   },
   "PXmagdtWJm98KndosY1UjC":{
      "tag1":2
   },
   "Mj3MhYW8YkDwTTLcLqbMX5":{
      "tag1":2
   },
   "VFUd8ynGaKB16okTwXmLMD":{
      "tag1":2
   },
   "36khQQPxy4zgafvD2LyVut":{
      "tag1":2
   },
   "J8RWLLj6dwPpHkTHrcpe8p":{
      "tag1":2
   },
   "PgPzVrwUKBrggyX1SjZkp1":{
      "tag1":2
   },
   "RQsksh7V4WnzJR7yqghk3U":{
      "tag1":3
   },
   "6AGJ3HHnyMwTEAHNfMmpC4":{
      "tag1":2
   },
   "3stYf4HSzz4wbH2SNzj4H4":{
      "tag1":3
   },
   "DpRvXmAnRxbdDbWwwfMsEN":{
      "tag1":2
   },
   "5nDyJVP1NrcPAttP3xwMB9":{
      "npdb":2
   },
   "9Mm2fqJGZxAocpCqkvJHBv":{
      "tag1":2
   },
   "T3z6hghzRH8L1HY4cRJZcG":{
      "training":2,
      "board_certification":2
   },
   "HR6vs6GEZ8rHaVgjg2WodM":{
      "tag":12
   },
   "7JRHtdHsxRhPFbrps8PudM":{
      "tag1":2
   },
   "QAV718LDSCXXyrG94WC4zo":{
      "tag1":7
   },
   "FJ9BcC4wjBgDkQo13aPDSL":{
      "tag2":2,
      "tag":2,
      "tag1":15
   },
   "QJKKQkReGbGkgbxVJTwvmt":{
      "mc":3
   },
   "VnnCsktkh8MvSvJ78pCSAw":{
      "mc":3
   },
   "5qDT9FeUYapqmZUrr91XNt":{
      "mc":2,
      "pmc":5
   },
   "Bmtdw641BTyB5pURHn7meX":{
      "mc":3
   },
   "RFHjUguGg8WbfqgeNjGf4x":{
      "mc":4
   },
   "51ZM2AeQKmggmSRDkpoE4S":{
      "tag1":2
   },
   "LGaV1RcYa4DukHg9Hk2XLX":{
      "tag1":2
   },
   "RV7pGxw45nKceoHYtkpysC":{
      "tag1":2
   },
   "R6kf9GCbVH3q536SB1HU9L":{
      "tag1":6
   },
   "EVm63ikhemr6zy5CfuRNgn":{
      "tag":3,
      "tag1":3
   },
   "UVRHMFxNsKhwQJLhbXANP":{
      "tag1":3
   },
   "MKKewEXvHrygeXhdQn2wm2":{
      "tag1":9
   },
   "78UtUwtSEdLhkKnesuE3bn":{
      "tag1":5
   },
   "AzL57UhUgequEjEGELwcVW":{
      "tag":2
   },
   "QR7qR9AiS8k8v7zqxyR5XL":{
      "tag1":2
   },
   "JNXZGR4Kfyakzi7Wu4Qics":{
      "tag":4
   },
   "HFfvmpjTcGkrTiWAw2WDnc":{
      "tag1":3
   },
   "RZj3oD5cjfp9wtn7YMWZ9c":{
      "tag":3
   }
}
```

**Comment by domwoe** [comment]
I thought about it a bit and I don't have a good solution...

One issue is in indy node. The key of the state trie is the current identifier which includes the `seqNo` of the referenced schema which would require either:
- An additional data structure/store to find a cred def without an explicit `seqNo` 
- Or we would need to migrate the state trie, which I don't think we want to do

So, maybe we should already introduce `v0` and `v1` identifiers:

- `v0` would allow identifying legacy cred defs and includes a `seqNo`
- `v1` are identified by a unique name (per issuer DID) only.

@dbluhm  Maybe you got a better idea?

**Comment by paulbastian** [comment]
Apart from this it seems weird to me that only Schema have a version and CredDef do not for a scalable ecosystem?

**Comment by domwoe** [comment]
yeah, I'd agree. This should probably be discussed in the Anoncreds working group. 

At first, I thought Indy node wouldn't need to care if we encode the name and version in the current tag. But it depends on where we want to enforce uniqueness: name or name + version.

If we only enforce name + version, it would open more possibilities to shoot oneself in the foot

**Comment by domwoe** [comment]
Closed with the merge of PR #49 in [Indy DID Method Specification Call](https://wiki.hyperledger.org/display/indy/2022-02-22%3A+Indy+DID+Method+Specification+Call)
